hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java

```diff
@@ -192,125 +166,173 @@ public void init(NodeEngine nodeEngine, Properties properties) {
+            Operation op = new ExplicitSuspicionOp(endpointMembersViewMetadata);
+            operationService.send(op, endpoint);
+        } else {
+            operationService.send(new MemberRemoveOperation(getThisAddress()), endpoint);
+        }
     }
 
-    void doRemoveAddress(Address deadAddress, String reason, boolean destroyConnection) {
-        if (!ensureMemberIsRemovable(deadAddress)) {
+    void sendExplicitSuspicionTrigger(Address triggerTo, MembersViewMetadata endpointMembersViewMetadata) {
+        if (triggerTo.equals(node.getThisAddress())) {
+            logger.warning("Cannot send explicit suspicion trigger for " + endpointMembersViewMetadata + " to itself.");
             return;
         }
 
+        int memberListVersion = membershipManager.getMemberListVersion();
+        Operation op = new TriggerExplicitSuspicionOp(memberListVersion, endpointMembersViewMetadata);
+        OperationService operationService = nodeEngine.getOperationService();
+        operationService.send(op, triggerTo);
+    }
+
+    public MembersView handleMastershipClaim(Address candidateAddress, String candidateUuid) {
+        // verify candidateAddress is not me DONE
+        // verify I am not master DONE
+        // verify candidateAddress is a valid member with its uuid
+        // verify I suspect everyone before the candidateAddress
+        // verify candidateAddress is not suspected.
+
+        checkNotNull(candidateAddress);
+        checkNotNull(candidateUuid);
+        checkFalse(getThisAddress().equals(candidateAddress), "cannot accept my own mastership claim!");
+
         lock.lock();
         try {
-            if (deadAddress.equals(node.getMasterAddress())) {
-                assignNewMaster();
-            }
-            if (node.isMaster()) {
-                clusterJoinManager.removeJoin(deadAddress);
+            checkTrue(isJoined(), candidateAddress + " claims mastership but this node is not joined!");
+            checkFalse(isMaster(),
+                    candidateAddress + " claims mastership but this node is master!");
+            checkFalse(candidateAddress.equals(getMasterAddress()),
+                    candidateAddress + " claims mastership but it is already the known master!");
+
+            MemberImpl masterCandidate = membershipManager.getMember(candidateAddress, candidateUuid);
+            checkTrue(masterCandidate != null ,
+                    candidateAddress + " claims mastership but it is not a member!");
+
+            MemberMap memberMap = membershipManager.getMemberMap();
+            if (!shouldAcceptMastership(memberMap, masterCandidate)) {
+                String message = "Cannot accept mastership claim of " + candidateAddress
+                        + " at the moment. There are more suitable master candidates in the member list.";
+                logger.fine(message);
+                throw new RetryableHazelcastException(message);
```

Comment:
What would happen if every node other than the one claiming mastership know that an older node is alive? E.g. We have [A,B,C,D,E]. Only B starts suspecting A and starts mastership claim. Other nodes will return `RetryableHazelcastException` to B since they know A is actually alive. B then receives a heartbeat from A but it will continue retrying the mastership claim.
Will B in the end become master only of itself - when the retry count is exceeded for all members? And then it will rejoin the cluster [A,C,D,E] by split brain handling?